File: python/src/developer_toolkit_elasticache/token_generator.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

import os
import re
from typing import TYPE_CHECKING
from urllib.parse import urlencode

from botocore.auth import SigV4QueryAuth
from botocore.awsrequest import AWSRequest
from botocore.session import Session

from developer_toolkit_elasticache.errors import (
    ConfigurationError,
    InvalidParameterError,
)

if TYPE_CHECKING:
    from botocore.credentials import ReadOnlyCredentials


_TOKEN_TTL_SECONDS = 900  # 15 minutes

_SIGNING_SERVICE = "elasticache"

# The signed request is a presigned URL; the token is that URL without the scheme.
_URL_SCHEME_PREFIX = "https://"

# The cache name is the SigV4 signing host, validating it keeps a malformed or
# hostile value from being signed against a host and producing a wrong token.
# Matching documented ElastiCache constraints in cache name.
_CACHE_NAME_PATTERN = re.compile(r"^[a-zA-Z][a-zA-Z0-9]*(-[a-zA-Z0-9]+)*$")

# ElastiCache service UserId pattern, with an optional "default." prefix for
# service-managed users (e.g. default.iam-user).
_USER_ID_PATTERN = re.compile(r"^(?:default\.)?[a-zA-Z][a-zA-Z0-9\-]*$")

# Honoured on top of botocore's own region resolution — see _resolve_region.
_REGION_ENV_VAR = "AWS_REGION"


_NO_REGION_MESSAGE = (
    "No AWS region found. Pass region explicitly, or configure one via "
    "AWS_REGION, AWS_DEFAULT_REGION, or the region setting in your AWS config profile."
)
_NO_CREDENTIALS_MESSAGE = (
    "No AWS credentials found. Configure credentials via the environment, "
    "shared config/credentials files, or an instance/container role."
)


def _validate_and_normalize(parameter: str, value: str, pattern: re.Pattern[str]) -> str:
    """Lowercase ``value``, check it against ``pattern``, and return the normalized form.

    ElastiCache stores cache names and user ids lowercase, so a value is normalized
    before both matching and signing; the value signed into the token and the value
    the client sends must match server-side.
    """
    normalized = value.lower()
    if not pattern.match(normalized):
        # Never echo the caller's value back in the message — name the parameter
        # and the accepted pattern only.
        raise InvalidParameterError(
            f"Invalid value for parameter {parameter!r}: must match {pattern.pattern}"
        )
    return normalized


def _resolve_target(
    serverless_cache_name: str | None, replication_group_id: str | None
) -> tuple[str, bool]:
    """Map the mutually-exclusive name arguments to ``(cache_name, serverless)``.
    The chosen value is validated here, under its own parameter name, so the
    error names the argument the caller passed.
    """
    if serverless_cache_name and not replication_group_id:
        cache_name = _validate_and_normalize(
            "serverless_cache_name", serverless_cache_name, _CACHE_NAME_PATTERN
        )
        return cache_name, True
    if replication_group_id and not serverless_cache_name:
        cache_name = _validate_and_normalize(
            "replication_group_id", replication_group_id, _CACHE_NAME_PATTERN
        )
        return cache_name, False
    raise InvalidParameterError(
        "Invalid parameter combination for 'serverless_cache_name' and "
        "'replication_group_id': exactly one must be provided."
    )


def _resolve_region(region: str | None, session: Session) -> str:
    """Fall back to a configured region when one was not passed explicitly.

    The session fallback is read from the same session that vends the credentials
    ``AWS_REGION`` is consulted directly because botocore does not: its region only
    resolves from ``AWS_DEFAULT_REGION`` or the config profile.
    """
    resolved = (
        region or os.environ.get(_REGION_ENV_VAR) or session.get_config_variable("region")
    )
    if not resolved:
        raise ConfigurationError(_NO_REGION_MESSAGE)
    return resolved


def _resolve_credentials(session: Session) -> "ReadOnlyCredentials":
    """Read a frozen credential set from the session's credential chain.

    Called on every token request rather than cached, so rotated credentials
    (SSO, assume-role, container, instance metadata) are picked up automatically.
    """
    credentials = session.get_credentials()
    if credentials is None:
        raise ConfigurationError(_NO_CREDENTIALS_MESSAGE)
    return credentials.get_frozen_credentials()


def _sign_token(
    cache_name: str,
    user_id: str,
    region: str,
    credentials: "ReadOnlyCredentials",
    *,
    serverless: bool = True,
) -> str:
    """Core signing logic. Returns the token string.

    ``cache_name`` is the ElastiCache **cache name** (for a serverless cache) or
    the **replication group id** (for a node-based cluster) — the value used as
    the SigV4 signing host. The server recomputes the signature using this name as
    the host.
    Cache names are lowercased at creation time, so the name must be
    signed in lowercase to avoid auth errors.

    ``user_id`` is signed as the ``User`` request parameter, lowercased because
    the service stores the user id lowercase (see ``_validate_and_normalize``).
    (For IAM-enabled users this must equal the Redis/Valkey ACL user name, but
    conceptually it is the ElastiCache user id, not the ACL name.)

    ``serverless`` is part of the signature, so it cannot be appended after signing.
    """
    user_id = _validate_and_normalize("user_id", user_id, _USER_ID_PATTERN)

    params = {"Action": "connect", "User": user_id}
    if serverless:
        params["ResourceType"] = "ServerlessCache"
    url = f"{_URL_SCHEME_PREFIX}{cache_name.lower()}/?{urlencode(params)}"

    request = AWSRequest(method="GET", url=url)
    signer = SigV4QueryAuth(
        credentials, _SIGNING_SERVICE, region, expires=_TOKEN_TTL_SECONDS
    )
    signer.add_auth(request)

    return request.url.removeprefix(_URL_SCHEME_PREFIX)


def generate_iam_auth_token(
    *,
    user_id: str,
    region: str | None = None,
    serverless_cache_name: str | None = None,
    replication_group_id: str | None = None,
    session: Session | None = None,
) -> str:
    """Generate an ElastiCache IAM auth token using the default AWS credential chain.

    Provide exactly one of ``serverless_cache_name`` (for a serverless cache) or
    ``replication_group_id`` (for a node-based cluster); the choice selects the
    resource type baked into the signature. ``user_id`` is the ElastiCache user
    id used to build the User ARN in the IAM policy. Returns the token string
    (like RDS's generate_db_auth_token).

    ``region`` defaults to the region the session resolves, so it only needs to be
    passed to sign for a region other than the configured one.

    Pass ``session`` to sign with a specific botocore session; by default a new
    session resolves credentials from the standard chain.
    """
    cache_name, serverless = _resolve_target(serverless_cache_name, replication_group_id)
    session = session or Session()
    resolved_region = _resolve_region(region, session)
    credentials = _resolve_credentials(session)
    return _sign_token(
        cache_name, user_id, resolved_region, credentials, serverless=serverless
    )


class ElastiCacheIAMAuthTokenProvider:
    """Generic IAM token provider for ElastiCache.

    Signs a fresh token on demand each time ``get_token()`` is called, then
    bridges it into whatever client you use (see the ``examples/`` directory).

    Provide exactly one of ``serverless_cache_name`` (for a serverless cache) or
    ``replication_group_id`` (for a node-based cluster).
    ``user_id`` is the ElastiCache user id.
    ``region`` defaults to the region the session resolves, so it only needs to be
    passed to sign for a region other than the configured one.

    Credentials come from the default AWS credential chain (env vars, shared
    config/credentials, or the instance/container role). They are re-read on
    every call, so rotated credentials (SSO, assume-role, container, instance
    metadata) are picked up automatically.
    """

    def __init__(
        self,
        *,
        user_id: str,
        region: str | None = None,
        serverless_cache_name: str | None = None,
        replication_group_id: str | None = None,
        session: Session | None = None,
    ):
        self._cache_name, self._serverless = _resolve_target(
            serverless_cache_name, replication_group_id
        )
        self._user_id = _validate_and_normalize("user_id", user_id, _USER_ID_PATTERN)
        self._session = session or Session()
        # Resolved once here rather than per call: unlike credentials, the region is
        # static configuration, and a missing region should surface at construction
        # instead of at the first connection attempt.
        self._region = _resolve_region(region, self._session)

    @property
    def user_id(self) -> str:
        return self._user_id

    def get_token(self) -> str:
        """Sign and return a fresh IAM auth token."""
        credentials = _resolve_credentials(self._session)
        return _sign_token(
            self._cache_name,
            self._user_id,
            self._region,
            credentials,
            serverless=self._serverless,
        )

File: node/src/token-generator.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { fromNodeProviderChain } from "@aws-sdk/credential-providers";
import { loadConfig, NODE_REGION_CONFIG_OPTIONS } from "@smithy/core/config";
import { HttpRequest } from "@smithy/core/protocols";
import { Hash } from "@smithy/hash-node";
import { SignatureV4 } from "@smithy/signature-v4";
import type {
  AwsCredentialIdentity,
  AwsCredentialIdentityProvider,
  QueryParameterBag,
} from "@smithy/types";

import { ConfigurationError, InvalidParameterError } from "./errors.js";

/** Lifetime of a presigned ElastiCache IAM token, in seconds. */
export const TOKEN_TTL_SECONDS = 900;
const SIGNING_SERVICE = "elasticache";

// The cache name becomes the SigV4 signing host. Keep this validation tight so a
// malformed or hostile value cannot be signed as a different host.
const CACHE_NAME_PATTERN = /^[a-zA-Z][a-zA-Z0-9]*(-[a-zA-Z0-9]+)*$/;

// ElastiCache service-managed user ids may use the special "default." prefix,
// for example "default.iam-user".
const USER_ID_PATTERN = /^(?:default\.)?[a-zA-Z][a-zA-Z0-9-]*$/;

const NO_REGION_MESSAGE =
  "No AWS region found. Pass region explicitly, or configure one via " +
  "AWS_REGION, AWS_DEFAULT_REGION, or the region setting in your AWS config profile.";
const NO_CREDENTIALS_MESSAGE =
  "No AWS credentials found. Configure credentials via the environment, " +
  "shared config/credentials files, or an instance/container role.";

export interface TokenGeneratorOptions {
  /** ElastiCache user id used in the signed `User` query parameter. */
  userId: string;
  /** AWS region. Resolution falls back to the supported AWS configuration sources. */
  region?: string;
  /** Name of a serverless cache. Exactly one target option is required. */
  serverlessCacheName?: string;
  /** Id of a node-based replication group. Exactly one target option is required. */
  replicationGroupId?: string;
}

/**
 * Advanced dependencies for deterministic tests and applications with an
 * explicitly managed AWS configuration. These are intentionally optional; the
 * normal API uses the standard Node.js AWS credential and region providers.
 */
export interface TokenGeneratorDependencies {
  credentialProvider?: AwsCredentialIdentityProvider;
  regionProvider?: () => Promise<string | undefined>;
  signingDate?: Date;
}

type Target = {
  cacheName: string;
  serverless: boolean;
};

type RegionResolution = { region: string } | { error: ConfigurationError };

function validateAndNormalize(
  parameter: string,
  value: unknown,
  pattern: RegExp,
): string {
  if (typeof value !== "string") {
    throw new InvalidParameterError(
      `Invalid value for parameter '${parameter}': must match ${pattern.source}`,
    );
  }

  // ElastiCache stores cache names and user ids lowercase server-side. Normalize
  // before validation and signing so the signed value matches the value sent by
  // the client; otherwise authentication can fail with WRONGPASS.
  const normalized = value.toLowerCase();
  if (!pattern.test(normalized)) {
    // Do not include the supplied value: it may contain newlines or URL syntax.
    throw new InvalidParameterError(
      `Invalid value for parameter '${parameter}': must match ${pattern.source}`,
    );
  }
  return normalized;
}

function resolveTarget(options: TokenGeneratorOptions): Target {
  const { serverlessCacheName, replicationGroupId } = options;

  // Validate inside each branch so an invalid value is reported under the public
  // parameter name the caller actually supplied.
  if (serverlessCacheName && !replicationGroupId) {
    return {
      cacheName: validateAndNormalize(
        "serverlessCacheName",
        serverlessCacheName,
        CACHE_NAME_PATTERN,
      ),
      serverless: true,
    };
  }

  if (replicationGroupId && !serverlessCacheName) {
    return {
      cacheName: validateAndNormalize(
        "replicationGroupId",
        replicationGroupId,
        CACHE_NAME_PATTERN,
      ),
      serverless: false,
    };
  }

  throw new InvalidParameterError(
    "Invalid parameter combination for 'serverlessCacheName' and " +
      "'replicationGroupId': exactly one must be provided.",
  );
}

async function resolveRegion(
  region: string | undefined,
  regionProvider: (() => Promise<string | undefined>) | undefined,
): Promise<string> {
  const fromArgument = region || undefined;
  const fromEnvironment = process.env.AWS_REGION || process.env.AWS_DEFAULT_REGION;
  if (fromArgument || fromEnvironment) {
    return fromArgument || fromEnvironment!;
  }

  // Environment variables were checked above; now fall back to the shared
  // AWS configuration while preserving the documented precedence.
  try {
    const resolved = await (regionProvider ?? loadSharedRegion)();
    if (resolved) {
      return resolved;
    }
  } catch {
    // Present one stable, actionable configuration error instead of leaking
    // provider internals or a file path from the local AWS configuration.
  }

  throw new ConfigurationError(NO_REGION_MESSAGE);
}

function retainRegion(
  region: string | undefined,
  regionProvider: (() => Promise<string | undefined>) | undefined,
): Promise<RegionResolution> {
  // Keep this promise fulfilled even when configuration is invalid. A provider
  // may be constructed without an immediate async error observer; getToken()
  // reports the retained configuration error when the token is requested.
  return resolveRegion(region, regionProvider).then(
    (resolvedRegion) => ({ region: resolvedRegion }),
    (error: unknown) => ({
      error:
        error instanceof ConfigurationError
          ? error
          : new ConfigurationError(NO_REGION_MESSAGE),
    }),
  );
}

async function loadSharedRegion(): Promise<string | undefined> {
  // Override the SDK selector/default so all "no region" cases return undefined
  // here and are converted by resolveRegion into one stable ConfigurationError,
  // rather than exposing the SDK's provider-specific error shape.
  const regionConfig = {
    ...NODE_REGION_CONFIG_OPTIONS,
    environmentVariableSelector: (env: Record<string, string | undefined>) =>
      env.AWS_REGION || env.AWS_DEFAULT_REGION,
    default: async () => undefined,
  };
  return loadConfig<string | undefined>(regionConfig)();
}

async function resolveCredentials(
  credentialProvider: AwsCredentialIdentityProvider,
): Promise<AwsCredentialIdentity> {
  try {
    // Resolve on every request instead of caching: SSO, assume-role, container,
    // and instance-metadata providers may rotate or refresh their credentials.
    const credentials = await credentialProvider();
    if (!credentials?.accessKeyId || !credentials.secretAccessKey) {
      throw new Error("The credential provider returned no usable credentials.");
    }
    return credentials;
  } catch {
    throw new ConfigurationError(NO_CREDENTIALS_MESSAGE);
  }
}

function encodeQueryComponent(value: string): string {
  // encodeURIComponent leaves !'()* unescaped, but SigV4 requires them to be
  // percent-encoded. Returning a different encoding from the one that was signed
  // makes the service canonicalize a different query and silently reject the token.
  return encodeURIComponent(value).replace(
    /[!'()*]/g,
    (character) => `%${character.charCodeAt(0).toString(16).toUpperCase()}`,
  );
}

function formatQuery(query: QueryParameterBag): string {
  // Preserve repeated query values while applying the SigV4-compatible encoding
  // above to every key and value in the final token.
  const entries: string[] = [];
  for (const [key, value] of Object.entries(query)) {
    if (value === null) {
      entries.push(`${encodeQueryComponent(key)}=`);
      continue;
    }
    const values = Array.isArray(value) ? value : [value];
    for (const item of values) {
      entries.push(`${encodeQueryComponent(key)}=${encodeQueryComponent(item)}`);
    }
  }
  return entries.join("&");
}

/**
 * Sign the ElastiCache IAM authentication request.
 *
 * `target.cacheName` is deliberately used as the hostname and `Host` header: it
 * is the SigV4 signing host expected by ElastiCache, not the connection endpoint
 * DNS name. For serverless caches, `ResourceType` is part of the canonical query,
 * so it must be added before presigning rather than appended afterward.
 */
async function signToken(
  target: Target,
  userId: string,
  region: string,
  credentials: AwsCredentialIdentity,
  signingDate: Date | undefined,
): Promise<string> {
  const query: QueryParameterBag = {
    Action: "connect",
    User: userId,
  };
  if (target.serverless) {
    // This parameter changes the canonical request and therefore the signature;
    // node-based replication groups must omit it.
    query.ResourceType = "ServerlessCache";
  }

  const request = new HttpRequest({
    protocol: "https:",
    hostname: target.cacheName,
    method: "GET",
    path: "/",
    headers: { host: target.cacheName },
    query,
  });
  const signer = new SignatureV4({
    credentials,
    region,
    service: SIGNING_SERVICE,
    sha256: Hash.bind(null, "sha256"),
  });
  const signed = await signer.presign(request, {
    expiresIn: TOKEN_TTL_SECONDS,
    signingDate,
  });

  return `${signed.hostname}/?${formatQuery(signed.query ?? {})}`;
}

function defaultCredentialProvider(): AwsCredentialIdentityProvider {
  return fromNodeProviderChain();
}

/**
 * Generate an ElastiCache IAM authentication token.
 *
 * Exactly one of `serverlessCacheName` or `replicationGroupId` must be provided.
 * The selected value is used as the SigV4 signing host; it is not the connection
 * endpoint DNS name. The `userId` is normalized to the lowercase value used in the
 * signed `User` query parameter.
 *
 * When `region` is omitted, resolution checks `AWS_REGION`, `AWS_DEFAULT_REGION`,
 * and the standard AWS shared configuration. Credentials use the standard Node.js
 * provider chain and are resolved for each invocation so refreshed credentials are
 * picked up.
 *
 * @param options Target, user, and optional region used to create the token.
 * @param dependencies Optional providers and signing time for controlled callers
 * and deterministic tests.
 * @returns The presigned ElastiCache IAM authentication token.
 */
export async function generateIamAuthToken(
  options: TokenGeneratorOptions,
  dependencies: TokenGeneratorDependencies = {},
): Promise<string> {
  const target = resolveTarget(options);
  const userId = validateAndNormalize("userId", options.userId, USER_ID_PATTERN);
  const region = await resolveRegion(options.region, dependencies.regionProvider);
  const credentials = await resolveCredentials(
    dependencies.credentialProvider ?? defaultCredentialProvider(),
  );

  return signToken(target, userId, region, credentials, dependencies.signingDate);
}

/**
 * Client-agnostic ElastiCache IAM token provider.
 *
 * The target and normalized user id are retained, and the region is resolved once
 * during construction. `getToken()` signs a fresh token and re-fetches credentials
 * on every call so rotated SSO, assume-role, container, or instance credentials are
 * used without changing the signing region.
 *
 * Region resolution is asynchronous, so a constructor cannot report a missing region.
 * Prefer `ElastiCacheIAMAuthTokenProvider.create()`, which awaits that resolution and
 * fails immediately; the constructor retains the error and reports it from the first
 * `getToken()` call instead.
 */
export class ElastiCacheIAMAuthTokenProvider {
  private readonly target: Target;
  private readonly userIdValue: string;
  private readonly credentialProvider: AwsCredentialIdentityProvider;
  private readonly regionResolution: Promise<RegionResolution>;
  private readonly signingDate?: Date;

  constructor(
    options: TokenGeneratorOptions,
    dependencies: TokenGeneratorDependencies = {},
  ) {
    this.target = resolveTarget(options);
    this.userIdValue = validateAndNormalize("userId", options.userId, USER_ID_PATTERN);
    this.credentialProvider =
      dependencies.credentialProvider ?? defaultCredentialProvider();
    this.regionResolution = retainRegion(options.region, dependencies.regionProvider);
    this.signingDate = dependencies.signingDate;
  }

  /**
   * Create a provider, resolving the region before returning.
   *
   * Use this instead of `new` to surface a missing or unreadable region
   * configuration up front rather than at the first connection attempt.
   *
   * @throws {InvalidParameterError} for an invalid target or user id.
   * @throws {ConfigurationError} when no region can be resolved.
   */
  static async create(
    options: TokenGeneratorOptions,
    dependencies: TokenGeneratorDependencies = {},
  ): Promise<ElastiCacheIAMAuthTokenProvider> {
    const provider = new ElastiCacheIAMAuthTokenProvider(options, dependencies);
    await provider.ensureRegionResolved();
    return provider;
  }

  /**
   * Await region resolution, throwing the retained configuration error if the
   * region could not be resolved.
   *
   * @throws {ConfigurationError} when no region can be resolved.
   */
  async ensureRegionResolved(): Promise<void> {
    const regionResolution = await this.regionResolution;
    if ("error" in regionResolution) {
      throw regionResolution.error;
    }
  }

  get userId(): string {
    return this.userIdValue;
  }

  /** Sign and return a fresh token using the provider's retained target and region. */
  async getToken(): Promise<string> {
    const regionResolution = await this.regionResolution;
    if ("error" in regionResolution) {
      throw regionResolution.error;
    }
    const credentials = await resolveCredentials(this.credentialProvider);
    return signToken(
      this.target,
      this.userIdValue,
      regionResolution.region,
      credentials,
      this.signingDate,
    );
  }
}
